jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java

```diff
@@ -15,23 +15,199 @@
+import org.eclipse.jetty.io.EofException;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.MethodSource;
 
+import static org.awaitility.Awaitility.await;
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.core.Is.is;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
 
 public class RequestReaderTest extends AbstractTest
 {
+    @ParameterizedTest
+    @MethodSource("transports")
+    public void testChannelStateSucceeded(Transport transport) throws Exception
+    {
+        CountDownLatch servletDoneLatch = new CountDownLatch(1);
+        start(transport, new HttpServlet()
+        {
+            @Override
+            protected void service(HttpServletRequest request, HttpServletResponse response) throws IOException
+            {
+                AsyncContext asyncContext = request.startAsync();
+
+                ServletInputStream inputStream = request.getInputStream();
+                inputStream.setReadListener(new ReadListener()
+                {
+                    @Override
+                    public void onDataAvailable() throws IOException
+                    {
+                        while (inputStream.isReady() && !inputStream.isFinished())
+                        {
+                            int read = inputStream.read();
+                            if (read < 0)
+                                break;
+                        }
+                    }
+
+                    @Override
+                    public void onAllDataRead()
+                    {
+                    }
+
+                    @Override
+                    public void onError(Throwable t)
+                    {
+                    }
+                });
+
+                response.sendError(567);
+
+                new Thread(() ->
+                {
+                    try
+                    {
+                        Thread.sleep(1000);
+                    }
+                    catch (InterruptedException e)
+                    {
+                        throw new RuntimeException(e);
+                    }
+                    finally
+                    {
+                        servletDoneLatch.countDown();
+                    }
+                }).start();
+            }
+        });
+
+        AtomicReference<Result> resultRef = new AtomicReference<>();
+        try (AsyncRequestContent content = new AsyncRequestContent())
+        {
+            Request request = client.newRequest(newURI(transport))
+                .method("POST")
+                .timeout(5, TimeUnit.SECONDS)
+                .body(content);
+            request.send(resultRef::set);
+            servletDoneLatch.await();
+            request.abort(new ArithmeticException());
```

Comment:
I don't think this line is relevant?

Let's try a bit harder to test that the server side `ServletChannel` state machine does not enter invalid states.
Can we notice the invalid states by registering listeners or similar?